### PR TITLE
Re-use XContentBuilder in Indexer

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Inspired by https://github.com/crate/crate/pull/13790

I tried once reusing XContentBuilder in CSVParser (https://github.com/crate/crate/pull/12795) and it gave some tiny improvement. Linked PR was closed as change was dependant on other components but here lifecycle of `XContentBuilder` seems to be isolated, we pass it to other places but close in the Indexer itself.

```
Benchmark                                             Mode  Cnt     Score    Error  Units
IndexerBenchmark.measure_index_baseline               avgt   10  1129,115 ±  6,927  us/op
IndexerBenchmark.measure_index_reuse_content_builder  avgt   10   901,766 ± 26,908  us/op
```

Changes in Benchmark and `reuse` flag are of course temporal - just did it in order to run all benchmarks in a single run.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java [hunk 1]
```diff
@@ -53,6 +53,19 @@ static BytesReference bytes(XContentBuilder xContentBuilder) {
         }
     }
 
+    /**
+     * Convert an {@link XContentBuilder} into a BytesReference.
+     * This method doesn't close the builder,so it can be re-used.
+     */
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java:59]: copy of `BytesReference.bytes` without `xContentBuilder.close()`. Tbh, I thought that closing builder is fine as even it closes stream it's safe - closing `BytesStreamOutput` is no-op anyway.

However, run into NPE when uses regular `bytes` call in this branch. Apparently it also closes internal generator.

File: server/src/main/java/io/crate/execution/dml/Indexer.java [hunk 2]
```diff
@@ -576,6 +591,9 @@ public ParsedDocument index(IndexItem item) throws IOException {
             );
         }
         xContentBuilder.endObject();
+        if (reuseContentBuilder) {
+            xContentBuilder.flush();
+        }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/dml/Indexer.java:596]: `BytesReference.bytes` is responsible for closing contentBuilder + underlying generator. 

Maybe need to add explicit `close` after the loop in the new version but not sure. Looked into `UTF8JsonGenerator.close`, it tries to end arrays/objects, flush which we did. Maybe need it to ensure  `_releaseBuffers` is called at the end.

**Review by mfussenegger — COMMENTED** [review]
Good idea :+1:

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java:59]: yep, actually don't even need a new method just do `stream.bytes`

File: server/src/main/java/io/crate/execution/dml/Indexer.java [hunk 3]
```diff
@@ -348,11 +349,13 @@ public Indexer(String indexName,
                    NodeContext nodeCtx,
                    Function<ColumnIdent, FieldType> getFieldType,
                    List<Reference> targetColumns,
-                   Symbol[] returnValues) {
+                   Symbol[] returnValues) throws IOException {
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/Indexer.java]: done in https://github.com/crate/crate/pull/13799/commits/7a59cf3822237a979c78d123b73231e958e19b00

File: server/src/main/java/io/crate/execution/dml/Indexer.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/Indexer.java]: What do we do with the reused XContentBuilder?  Closing BytesOutoutStream is defo not needed but stuff inside `_releaseBuffers` bother me:
`_ioContext.releaseWriteEncodingBuffer(buf) -> _bufferRecycler.releaseByteBuffer(BufferRecycler.BYTE_WRITE_ENCODING_BUFFER, buf);`

But if we close it we need to make Indexer closable and find out where is the end of its lifecycle - could you point out to that?

**Comment by BaurzhanSakhariev** [comment]
Some tests are failing:

1. test_fields_order_in_source_is_determinisitc

`fields in  {"x":10,"o":{"IGWITbgWNcZSiP":0,"wOSJCa":1,"GMcwNhbhetVfe":2,"ErtJXu":3,"hsnegKjadb":4,"MXzinJIuBf":5,"hTzxNvq":6},"y":50} must have same order in  {"x":10,"o":{"IGWITbgWNcZSiP":0,"wOSJCa":1,"GMcwNhbhetVfe":2,"ErtJXu":3,"hsnegKjadb":4,"MXzinJIuBf":5,"hTzxNvq":6},"y":50] `

2. .test_index_fails_if_not_null_column_has_null_value
Can not start an object, expecting field name (context: Object)

Feel like all this can be mitigated by carefully checking null values and don't do start/endObject but this is not a quick improvement anymore. needs a closer look. closing it for now

**Pull request closed without merging**